packages/core/functions/src/services/function-invocation-service.ts

```diff
@@ -0,0 +1,70 @@
+//
+// Copyright 2025 DXOS.org
+//
+import { Context, Effect, Layer } from 'effect';
+
+import { AiService } from '@dxos/ai';
+
+import { type FunctionDefinition } from '../handler';
+
+import { CredentialsService } from './credentials';
+import { DatabaseService } from './database';
+import {
+  FunctionImplementationResolver,
+  type InvocationServices,
+  LocalFunctionExecutionService,
+} from './local-function-execution';
+import { QueueService } from './queues';
+import { RemoteFunctionExecutionService } from './remote-function-execution-service';
+
+export class FunctionInvocationService extends Context.Tag('@dxos/functions/FunctionInvocationService')<
+  FunctionInvocationService,
+  {
+    invokeFunction<I, O>(
+      functionDef: FunctionDefinition<I, O>,
+      input: I,
+      deployedFunctionId?: string,
```

Comment:
I think this API isn't great -- the caller now needs to know if the function is an EDGE function or a deployed function.